src/System.Private.CoreLib/src/Internal/Runtime/Augments/EnvironmentAugments.cs

```diff
@@ -12,84 +13,169 @@ namespace Internal.Runtime.Augments
+            SetEnvironmentVariableCore(variable, value);
         }
 
-        private static int s_latchedExitCode;
-        public static int ExitCode
+        private static void ValidateVariableAndValue(string variable, ref string value)
         {
-            get
+            const int MaxEnvVariableValueLength = 32767;
+
+            if (variable == null)
+                throw new ArgumentNullException(nameof(variable));
+
+            if (variable.Length == 0)
+                throw new ArgumentException(SR.Argument_StringZeroLength, nameof(variable));
+
+            if (variable[0] == '\0')
+                throw new ArgumentException(SR.Argument_StringFirstCharIsZero, nameof(variable));
+
+            if (variable.Length >= MaxEnvVariableValueLength)
+                throw new ArgumentException(SR.Argument_LongEnvVarValue, nameof(variable));
+
+            if (variable.IndexOf('=') != -1)
+                throw new ArgumentException(SR.Argument_IllegalEnvVarName, nameof(variable));
+
+            if (string.IsNullOrEmpty(value) || value[0] == '\0')
             {
-                return s_latchedExitCode;
+                // Explicitly null out value if it's empty
+                value = null;
             }
-            set
+            else if (value.Length >= MaxEnvVariableValueLength)
             {
-                s_latchedExitCode = value;
+                throw new ArgumentException(SR.Argument_LongEnvVarValue, nameof(value));
             }
         }
 
-        private static string[] s_commandLineArgs;
-
-        internal static void SetCommandLineArgs(string[] args)
+        // TODO Perf: Once CoreCLR gets PopulateEnvironmentVariables(), get rid of GetEnvironmentVariables() and have 
+        // corefx call PopulateEnvironmentVariables() instead so we don't have to create a dictionary just to copy it into
+        // another dictionary.
+        public static IDictionary GetEnvironmentVariables()
         {
-            s_commandLineArgs = args;
+            IDictionary dictionary = new Dictionary<object, object>();
+            PopulateEnvironmentVariables(dictionary);
+            return dictionary;
         }
 
-        public static string[] GetCommandLineArgs()
+        public static void PopulateEnvironmentVariables(IDictionary dictionary)
         {
-            return (string[])s_commandLineArgs?.Clone();
+            unsafe
+            {
+                char* unsafeBlock = GetEnvironmentStrings();
+                if (unsafeBlock == (char*)0)
+                    throw new OutOfMemoryException();
+                try
+                {
+                    // Format for GetEnvironmentStrings is:
+                    // [=HiddenVar=value\0]* [Variable=value\0]* \0
+                    // See the description of Environment Blocks in MSDN's
+                    // CreateProcess page (null-terminated array of null-terminated strings).
+
+                    // Search for terminating \0\0 (two unicode \0's).
```

Comment:
The CoreCLR approach is to marshal the Unix EB into a Windows-style EB in the PAL and share all the managed approach. Since the eventual aim is to get this in the shared partition, that's assumed approach I followed. Let's discuss in the main CR that I'm about to put up.